src/cargo/core/compiler/custom_build.rs

```diff
@@ -429,6 +447,29 @@ fn build_work<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoRes
     Ok(job)
 }
 
+fn insert_warnings_in_build_output(
+    build_script_outputs: Arc<Mutex<BuildScriptOutputs>>,
+    id: PackageId,
+    metadata_hash: Metadata,
+    warnings: Vec<String>,
+) {
+    let build_output_with_only_warnings = BuildOutput {
+        warnings,
+        library_paths: Vec::new(),
+        library_links: Vec::new(),
+        linker_args: Vec::new(),
+        cfgs: Vec::new(),
+        env: Vec::new(),
+        metadata: Vec::new(),
+        rerun_if_changed: Vec::new(),
+        rerun_if_env_changed: Vec::new(),
```

Comment:
Maybe derive Default on BuildOutput, and then use default() here to avoid specifying all these fields.